File: yt/yt/flow/library/cpp/pipeline_tables/definitions.cpp
#include "public.h"

#include <yt/yt/client/table_client/schema.h>

#include <yt/yt/core/ytree/attributes.h>
#include <yt/yt/core/ytree/convert.h>
#include <yt/yt/core/ytree/node.h>

#include <library/cpp/resource/resource.h>

#include <utility>

namespace NYT::NFlow {

using namespace NTableClient;
using namespace NYTree;

////////////////////////////////////////////////////////////////////////////////

namespace {

constexpr TStringBuf DefinitionsResourceKey =
    "resfs/file/yt/yt/flow/library/pipeline_tables/definitions.yson";

std::map<std::string, TPipelineTableDefinition> ParseSection(
    const IMapNodePtr& root,
    TStringBuf section)
{
    std::map<std::string, TPipelineTableDefinition> result;
    for (const auto& [name, descriptorNode] : root->GetChildOrThrow(section)->AsMap()->GetChildren()) {
        auto schemaNode = descriptorNode->AsMap()->GetChildOrThrow("schema");
        auto schemaYson = NYson::ConvertToYsonString(schemaNode);
        TPipelineTableDefinition definition{
            .Schema = ConvertTo<TTableSchemaPtr>(schemaYson),
            .SchemaYson = std::move(schemaYson),
            .Attributes = IAttributeDictionary::FromMap(
                descriptorNode->AsMap()->GetChildOrThrow("attributes")->AsMap()),
        };
        result.emplace(name, std::move(definition));
    }
    return result;
}

} // namespace

////////////////////////////////////////////////////////////////////////////////

const TPipelineTableDefinitions& GetPipelineTableDefinitions()
{
    static const auto Definitions = [] {
        auto root = ConvertTo<INodePtr>(NYson::TYsonString(NResource::Find(DefinitionsResourceKey)))->AsMap();
        return TPipelineTableDefinitions{
            .Tables = ParseSection(root, "tables"),
            .Queues = ParseSection(root, "queues"),
        };
    }();
    return Definitions;
}

IAttributeDictionaryPtr BuildPipelineTableAttributes(
    const TPipelineTableDefinition& definition)
{
    auto attributes = definition.Attributes->Clone();
    attributes->SetYson("schema", definition.SchemaYson);
    return attributes;
}

////////////////////////////////////////////////////////////////////////////////

} // namespace NYT::NFlow

File: yt/yt/flow/library/python/pipeline_tables/definitions.py
from copy import deepcopy
from pathlib import Path

import yt.yson as yson

try:
    from library.python import resource as _resource
except ModuleNotFoundError:
    _resource = None

_RESOURCE_KEY = "resfs/file/yt/yt/flow/library/pipeline_tables/definitions.yson"


def _load_section(definitions, section):
    return {
        str(name): {
            "schema": deepcopy(descriptor["schema"]),
            "attributes": deepcopy(descriptor["attributes"]),
        }
        for name, descriptor in definitions[section].items()
    }


def _schemas_only(definitions):
    return {name: {"schema": deepcopy(descriptor["schema"])} for name, descriptor in definitions.items()}


def _load_definitions():
    if _resource is not None:
        return yson.loads(_resource.find(_RESOURCE_KEY))

    try:
        data = Path(__file__).with_name("definitions.yson").read_bytes()
    except FileNotFoundError:
        source_path = Path(__file__).resolve().parents[2] / "pipeline_tables" / "definitions.yson"
        data = source_path.read_bytes()
    return yson.loads(data)


def _get_pipeline_table_definitions():
    return (
        _load_section(_DEFINITIONS, "tables"),
        _load_section(_DEFINITIONS, "queues"),
    )


_DEFINITIONS = _load_definitions()

PIPELINE_FILES = []
PIPELINE_TABLES = _schemas_only(_DEFINITIONS["tables"])
PIPELINE_QUEUES = _schemas_only(_DEFINITIONS["queues"])
